crema/qvalues.py

```diff
@@ -132,3 +136,220 @@ def _fdr2qvalue(scores, fdr):
+        # Create an array of bootstrapped p-values, and sort in ascending order.
+        num_draw = min(n_pval, max_size)
+        pBoot_list = np.random.choice(pval_list, size=num_draw, replace=True)
+        pBoot_list.sort()
+
+        for idx in range(0, len(lambda_list)):
+            start = np.searchsorted(pBoot_list, lambda_list[idx])
+            W1 = num_draw - start
+            pi0Boot = W1 / num_draw / (1.0 - lambda_list[idx])
+
+            # Estimated mean-squared error
+            mse_list[idx] += (pi0Boot - minPi0) * (pi0Boot - minPi0)
+
+    # Which index did the iterator go?
+    minIdx = np.argmin(mse_list)
+
+    # LOGGER.debug(f"Estimated pi0=%f at lambda=%f (MSE=%f)", pi0s_list[minIdx], lambda_list[minIdx], mse_list[minIdx])
+
+    pi0 = max(min(pi0s_list[minIdx], 1.0), 0.0)
+    return pi0
+
+
+@nb.njit
+def calculate_mixmax_qval(target_scores, decoy_scores, pi0):
+    """
+    Estimate q-values using mix-max.
+    """
+    # Note that the notation in this function follows the notation found in
+    # Percolator, which itself follows the notation found in Supplementary Note
+    # 3 (Keich et al., JPR. 2015.). This supplment can be found at
+    # http://dx.doi.org/10.1021/acs.jproteome.5b00081.
+
+    # assert pi0 >= 0 and pi0 < 1
+
+    num_targets = target_scores.shape[0]
+    num_decoys = decoy_scores.shape[0]
+
+    h_w_le_z = np.zeros(num_decoys + 1)  # histogram for N_{w<=z}
+    h_z_le_z = np.zeros(num_decoys + 1)  # histogram for N_{z<=z}
+
+    for i in range(0, num_decoys):
+        h_w_le_z[i] = np.searchsorted(
+            target_scores, decoy_scores[i], side="right"
+        )
+        h_z_le_z[i] = np.searchsorted(
+            decoy_scores, decoy_scores[i], side="right"
+        )
+
+    h_w_le_z[num_decoys] = num_targets
+    h_z_le_z[num_decoys] = num_decoys
+
+    fdrmod = np.zeros(num_targets)
+    estPx_lt_zj = 0.0
+    E_f1_mod_run_tot = 0.0
+    j = num_decoys - 1
+    k = num_targets - 1
+    n_z_ge_w = 0
+    n_w_ge_w = 0
+    prev_fdr = -1
+
+    for i in range(num_targets - 1, -1, -1):
+        while j >= 0 and decoy_scores[j] >= target_scores[i]:
+            cnt_w = h_w_le_z[j + 1]
+            cnt_z = h_z_le_z[j + 1]
+
+            estPx_lt_zj = (cnt_w - pi0 * cnt_z) / ((1.0 - pi0) * cnt_z)
+            if estPx_lt_zj > 1:
+                estPx_lt_zj = 1.0
+            elif estPx_lt_zj < 0:
+                estPx_lt_zj = 0.0
+
+            E_f1_mod_run_tot += estPx_lt_zj * (1.0 - pi0)
+            n_z_ge_w += 1
+            j -= 1
+
+        # not sure if this faster or if original while loop is faster
+        # AssignConfidence.cpp:1225
+        n_w_ge_w = (target_scores >= target_scores[i]).sum()
+        qvalue = (n_z_ge_w * pi0 + E_f1_mod_run_tot) / n_w_ge_w
+
+        if qvalue > 1:
+            qvalue = 1.0
+        fdrmod[i] = qvalue
+
+        # convert qvalues to fdr
+        if prev_fdr > fdrmod[i]:
+            fdrmod[i] = prev_fdr
+        prev_fdr = fdrmod[i]
```

Comment:
This logic gives incorrect _q_-values which are larger than those from `_fdr2qvalue` above. It results in large _q_-values being propagated to more relaxed score cutoffs, when the correct implementation propagates small _q_-values to stricter score cutoffs.

Recall that the _q_-value for a hypothesis is the minimum FDR for any subset of the list containing that hypothesis. Because this loop proceeds from strict to relaxed thresholds a separate loop in the opposite direction is required to compute the _q_-value.

Reply:
I will provide a PR patching this and the off-by-one bug noted above.

Reply:
Thanks again. I'll take a look. 
Note to self. Check mix-max implementation in Crux as well.

Reply:
I can confirm that Seth is correct regarding the incorrect q-values.
To do. check one off error.
Will create bug report in Crux code.